surface/include/pcl/surface/impl/mls.hpp

```diff
@@ -673,45 +534,363 @@ pcl::MovingLeastSquares<PointInT, PointOutT>::performUpsampling (PointCloudOut &
+
+      if (vi >= 1)
+        d.z_v += mls_result.c_vec[j] * vi * u_pow(ui) * v_pow(vi - 1);
+
+      if (ui >= 1 && vi >= 1)
+        d.z_uv += mls_result.c_vec[j] * ui * u_pow(ui - 1) * vi * v_pow(vi - 1);
+
+      if (ui >= 2)
+        d.z_uu += mls_result.c_vec[j] * ui * (ui - 1) * u_pow(ui - 2) * v_pow(vi);
+
+      if (vi >= 2)
+        d.z_vv += mls_result.c_vec[j] * vi * (vi - 1) * u_pow(ui) * v_pow(vi - 2);
+
+      if (ui == 0)
+        v_pow(vi + 1) = v_pow(vi) * v;
+
+      ++j;
+    }
+    u_pow(ui + 1) = u_pow(ui) * u;
+  }
+
+  return d;
+}
+
+Eigen::Vector2f pcl::calculatePrincipleCurvatures (const double u, const double v, const pcl::MLSResult &mls_result)
+{
+  Eigen::Vector2f k;
+
+  k << MLS_MINIMUM_PRINCIPLE_CURVATURE, MLS_MINIMUM_PRINCIPLE_CURVATURE;
+  // Note: this use the Monge Patch to derive the Gaussian curvature and Mean Curvature found here http://mathworld.wolfram.com/MongePatch.html
+  // Then:
+  //      k1 = H + sqrt(H^2 - K)
+  //      k1 = H - sqrt(H^2 - K)
+  if (mls_result.polynomial_fit && mls_result.c_vec.size () >= (mls_result.order + 1) * (mls_result.order + 2) / 2 && pcl_isfinite (mls_result.c_vec[0]))
+  {
+    PolynomialPartialDerivative d = getPolynomialPartialDerivative (u, v, mls_result);
+    double Z = 1 + d.z_u * d.z_u + d.z_v * d.z_v;
+    double Zlen = std::sqrt (Z);
+    double K = (d.z_uu * d.z_vv - d.z_uv * d.z_uv) / (Z * Z);
+    double H = ((1.0 + d.z_v * d.z_v) * d.z_uu - 2.0 * d.z_u * d.z_v * d.z_uv + (1.0 + d.z_u * d.z_u) * d.z_vv) / (2.0 * Zlen * Zlen * Zlen);
+    double disc2 = H * H - K;
+    assert (disc2 >= 0.0);
+    double disc = std::sqrt(disc2);
+    k[0] = H + disc;
+    k[1] = H - disc;
+
+    if (std::abs (k[0]) > std::abs (k[1])) std::swap (k[0], k[1]);
+  }
+  else
+  {
+    PCL_ERROR("No Polynomial fit data, unable to calculate the principle curvatures!\n");
+  }
+
+  return k;
+}
+
+pcl::MLSProjectionResults pcl::projectPointOrthogonalToPolynomialSurface (const double u, const double v, const double w, const pcl::MLSResult &mls_result)
+{
+  double gu = u;
+  double gv = v;
+  double gw = 0;
+
+  MLSProjectionResults result;
+  result.normal = mls_result.plane_normal;
+  if (mls_result.polynomial_fit && mls_result.c_vec.size () >= (mls_result.order + 1) * (mls_result.order + 2) / 2 && pcl_isfinite (mls_result.c_vec[0]))
+  {
+    PolynomialPartialDerivative d = getPolynomialPartialDerivative (gu, gv, mls_result);
+    gw = d.z;
+    double err_total;
+    double dist1 = std::abs (gw - w);
+    double dist2;
+    do
+    {
+      double e1 = (gu - u) + d.z_u * gw - d.z_u * w;
+      double e2 = (gv - v) + d.z_v * gw - d.z_v * w;
+
+      double F1u = 1 + d.z_uu * gw + d.z_u * d.z_u - d.z_uu * w;
+      double F1v = d.z_uv * gw + d.z_u * d.z_v - d.z_uv * w;
+
+      double F2u = d.z_uv * gw + d.z_v * d.z_u - d.z_uv * w;
+      double F2v = 1 + d.z_vv * gw + d.z_v * d.z_v - d.z_vv * w;
+
+      Eigen::MatrixXd J (2, 2);
```

Comment:
This matrix has known size at compile time.

Reply:
Yes.

Reply:
I think the idea behind Sergio's message was that `J` can be changed to a fixed-sized Eigen type, which will (maybe) make the following matrix inverse computation faster by using a specialized 2x2 implementation (compared to generic size implementation).

Actually, I would expect Eigen to be smart enough to perform runtime dispatching to specialized implementations, but we'd need to dig into the sources to find this out. So the easiest way to ensure we are as efficient as possible is to just switch to `Eigen::Matrix2d`.

Reply:
>  I think the idea behind Sergio's message was that J can be changed to a fixed-sized Eigen type, which will (maybe) make the following matrix inverse computation faster by using a specialized 2x2 implementation (compared to generic size implementation).

It was this exactly. Sorry for being unclear.